### DISCUSSION TITLE
Multi-file Vue components and $emit

### DISCUSSION THREAD

**Discussion by jazl** [widgetti/solara#216] [Q&A]
I am trying to use a Vuetify component that consists of 2 files, Parent.vue and Child.vue (with the Child component being imported into the Parent component).
[See code here](https://codesandbox.io/embed/vuetify-test-forked-4fpsxj?fontsize=14&hidenavigation=1&theme=dark)

Here is what it looks like running under Vue:
![image](https://github.com/widgetti/solara/assets/2700780/4c041144-2a71-44db-a906-a32c95fe05a1)

However, when trying to use Parent component in Solara, I get the following error:
>ReferenceError: Child is not defined
    at eval (eval at <anonymous> (VueTemplateRenderer.js:346:21), <anonymous>:6:5)
    at VueTemplateRenderer.js:346:21
    at or (VueTemplateRenderer.js:47:19)
    at VueTemplateRenderer.js:27:24
    at mr (VueRenderer.js:98:12)
    at VueRenderer.js:282:16
    at Array.map (<anonymous>)
    at Or (VueRenderer.js:271:36)
    at o.render (VueRenderer.js:118:111)
    at e._render (vue.runtime.esm.js:3569:22)

Here is my Solara code:
```
@solara.component_vue("./vue/Parent.vue")
def Parent():
    pass

@solara.component
def Page():
    Parent()
```
Am I missing some steps here? Also, how do I wire up the ‘vue-event’ event, which is $emitted by the Parent component on button click, on the Solara side?

**Comment by jazl** [comment]
I also tried looking at the unit tests but the Vue file is empty?
https://github.com/widgetti/solara/blob/master/tests/unit/component_vue_test.vue

**Comment by mariobuikhuizen** [comment]
The idea of Solara is that you can use the Python API without having to resort to front-end development. If you need something custom that is not available in the Python API, you can fall back on using some frontend code by using a VuetifyTemplate (which you are using in your example via the `@component_vue` decorator): https://ipyvuetify.readthedocs.io/en/latest/template_usage.html, https://solara.dev/examples/general/vue_component. You can embed Solara components in those templates if needed; see: https://github.com/widgetti/solara/blob/master/solara/components/download.vue and https://github.com/widgetti/solara/blob/master/solara/components/file_download.py.

If you have a large template or you have some template code you want to reuse in multiple templates, you can use the still undocumented "full-vue component"; see: https://github.com/widgetti/ipyvue/blob/master/examples/fullVueComponent.ipynb, but we rarely use this method ourselves.

Seeing the codesandbox example, I think you are trying to develop too much like is done in Vue. Solara uses Vue as an implementation detail to implement the frontend and offers an escape hatch to vue so users can extend Solara with frontend code, but it has a different model on how to structure your code, which should be mainly done in Python.

**Reply by jazl** [reply]
Thanks Mario. Will review the info you provided soon!

**Comment by mariobuikhuizen** [comment]
The example would look like something like this in Solara: (it can be run in a notebook cell)
```python
import solara

@solara.component
def Child():
    solara.Div(
        children=["Hello from Child component"],
        style="border: 2px green solid; padding: 10px; background-color: lightgray;")

    
@solara.component
def Parent(on_message):
    counter, set_counter = solara.use_state(0)
        
    with solara.Div(classes=["parent"]):
        Child()
        solara.Button(
            "say hello",
            on_click=lambda: on_message("Hello from Parent component"),
            color="primary")
        solara.Button(
            f"increment counter {counter}",
            on_click=lambda: set_counter(counter + 1),
            color="yellow")
        solara.Style("""
            .parent {
                border: 2px red solid;
                padding: 10px;
            }
            .parent button {
                margin: 4px;
            }
            """)
        
        
@solara.component
def App():
    message, set_message = solara.use_state("")
    
    with solara.v.Sheet():
        Parent(set_message)
        solara.Text(f"Got value: {message}")

App()
```
